src/Engine/ProtoCore/Reflection/MemberInfoReflectionExtensions.cs

```diff
@@ -0,0 +1,208 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Text;
+using System.Threading.Tasks;
+using Autodesk.DesignScript.Runtime;
+
+namespace ProtoCore.Reflection
+{
+    internal static class MemberInfoReflectionExtensions
+    {
+        internal static MethodInfo GetBaseDefinitionReflectionContext(this MethodInfo methodInfo)
+        {
+            try
+            {
+                // if overriding object 
+                var baseType = methodInfo.DeclaringType.BaseType;
+                if (baseType is null)
+                {
+                    return methodInfo;
+                }
+
+                // This will fail if there are overloads of the method
+                var baseTypeMethod = baseType.GetMethod(methodInfo.Name);
+
+                if (baseTypeMethod is null)
+                {
+                    return methodInfo;
+                }
+
+                // Need to return the original methodInfo if the method is hiding the inherited member using the new keyword
+                // This is probably not a supported way of doing this... but getting desperate
+                if (baseTypeMethod.IsVirtual != true && baseTypeMethod.GetMethodBody() != null)
+                {
+                    return methodInfo;
+                }
+
+                return baseTypeMethod;
+            }
+            catch (Exception)
+            {
+                return null;
+            }
+
+        }
+
+
+        internal static Attribute[] GetAttributesFromReflectionContext(this MemberInfo member)
+        {
+            var customAttributes = CustomAttributeData.GetCustomAttributes(member);
+            return GetReflectionAttributes(customAttributes);
+        }
+
+        internal static Attribute[] GetAttributesFromReflectionContext(this ParameterInfo parameter)
+        {
+            var customAttributes = CustomAttributeData.GetCustomAttributes(parameter);
+            return GetReflectionAttributes(customAttributes);
+        }
+
+        internal static Attribute[] GetAttributesFromReflectionContext(this System.Type type)
+        {
+            var customAttributes = CustomAttributeData.GetCustomAttributes(type);
+            return GetReflectionAttributes(customAttributes);
+        }
+
+        private static Attribute[] GetReflectionAttributes(IList<CustomAttributeData> customAttributes)
+        {
+            var attributes = new List<Attribute>();
+
+            foreach (var ca in customAttributes)
+            {
+                try
+                {
+                    var type = ca.AttributeType;
+
+                    var paramValues = new List<dynamic>();
+                    for (int i = 0; i < ca.ConstructorArguments.Count; i++)
```

Comment:
So @SHKnudsen - my interpretation of this code is a ton of work trying to get the attributes and their parameters and create instances of those parameters and attributes - but why?

Reply:
It does make me think that there might be a way to do this without using reflection at all - considering that the only attributes we care about will always be loaded as part of Dynamo's core binaries.

Maybe we can talk about this approach some more?

Reply:
Yes this is used so that we can easily use the same Attribute checks that are done in the CLRModuleType which relies on the actual attribute type. When in reflection we can only get CustomAttributeData types so here im using that data to recreate the actual attribute type

Reply:
> It does make me think that there might be a way to do this without using reflection at all - considering that the only attributes we care about will always be loaded as part of Dynamo's core binaries.

happy to chat more on this and hear your ideas